lib/minfraud/transaction.rb

```diff
@@ -28,60 +32,139 @@ class Transaction
     attr_accessor :avs_result, :cvv_result
 
     # Miscellaneous attribute (optional)
-    attr_accessor :requested_type, :forwarded_ip
-
+    attr_accessor :requested_type
+
+    ATTRIBUTES = [
+      :ip, :txn_id, :license_key,
+      :city, :state, :postal, :country,
+      :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country,
+      :email_domain, :email_md5, :phone,
+      :bin,
+      :user_agent, :accept_language,
+      :amount, :currency, :txn_type, :shop_id,
+      :avs_result, :cvv_result,
+      :requested_type
+    ].freeze
+
+    INPUT_STRING_ATTRIBUTES = [
+      :ip, :txn_id, :license_key,
+      :city, :state, :postal, :country,
+      :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country,
+      :email, :phone,
+      :bin,
+      :user_agent, :accept_language,
+      :currency, :txn_type,
+      :avs_result, :cvv_result,
+      :requested_type
+    ].freeze
+
+    INPUT_NUMERIC_ATTRIBUTES = [:amount].freeze
+
+    REQUIRED_ATTRIBUTES = [:ip, :txn_id, :license_key].freeze
+
+    CONVERT_STRING_ATTRIBUTES = [:shop_id].freeze
+
+    # Initializes the Transaction using parameters set from the block
+    # @raise [TransactionError] if parameters are set incorrectly
     def initialize
       yield self
       unless has_required_attributes?
         raise TransactionError, 'You did not set all the required transaction attributes.'
       end
       validate_attributes
+      freeze
     end
 
     # Hash of attributes that have been set
     # @return [Hash] present attributes
     def attributes
-      attrs = [:ip, :city, :state, :postal, :country, :license_key, :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country, :email_domain, :email_md5, :phone, :bin, :session_id, :user_agent, :accept_language, :txn_id, :amount, :currency, :txn_type, :avs_result, :cvv_result, :requested_type, :forwarded_ip, :shop_id]
-      attrs.map! do |a|
-        [a, send(a)]
-      end
-      Hash[attrs]
+      Hash[ATTRIBUTES.map { |a| [a, send(a)] }].compact
     end
 
     private
 
     # Ensures the required attributes are present
     # @return [Boolean]
     def has_required_attributes?
-      ip && txn_id && license_key
+      REQUIRED_ATTRIBUTES.none? { |attr| send(attr).nil? || send(attr).empty? }
     end
 
     # Validates the types of the attributes
-    # @return [nil, TransactionError]
+    # @raise [TransactionError] if present attributes are not valid
+    # @return [nil]
     def validate_attributes
-      [:ip, :city, :state, :postal, :country, :txn_id, :license_key].each { |s| validate_string(s) }
+      INPUT_STRING_ATTRIBUTES.each { |attr| validate_string(attr) }
+      INPUT_NUMERIC_ATTRIBUTES.each { |attr| validate_number(attr) }
+      CONVERT_STRING_ATTRIBUTES.each { |attr| convert_to_string(attr) }
+      validate_cvv
```

Comment:
Isn't `cvv_result` already being validated by validating `INPUT_STRING_ATTRIBUTES`?

Reply:
That's an interesting assumption - `INPUT_STRING_ATTRIBUTES` only validates that the attributes are strings, without any additional checks.

The `cvv_result` is one attribute that is being validated more strictly (it should be a single character), and more checks may be added in the future.

Should I add an array of `SINGLE_CHARACTER_ATTRIBUTES` and verify in a similar manner instead of doing this with `validate_cvv`?